File: events/meetup_dotnet_architects_2015/Fun with Functional Programming/C#/ListFuncs.cs
//Sample provided by Fabio Galuppo 
//February 2015 

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace CSharpFPIntro
{
    public static class EnumerableExtension
    {
        public static T Head<T>(this IEnumerable<T> xs)
        {
            return xs.First();
        }

        public static IEnumerable<T> Tail<T>(this IEnumerable<T> xs)
        {
            return xs.Skip(1);
        }

        public static IEnumerable<T> Init<T>(this IEnumerable<T> xs)
        {
            IEnumerator<T> e = xs.GetEnumerator();
            bool cont = e.MoveNext();
            while (cont)
            {
                var x = e.Current;
                cont = e.MoveNext();
                if (cont)
                    yield return x;
            }
        }

        #region "MakeString"
        public static string MakeString<T>(this IEnumerable<T> xs)
        {
            StringBuilder sb = new StringBuilder();
            foreach (var x in xs) sb.Append(x);
            return sb.ToString();
        }

        public static string MakeString<T>(this IEnumerable<T> xs, string sep)
        {
            StringBuilder sb = new StringBuilder();
            foreach (var x in xs.Init()) sb.AppendFormat("{0}{1}", x, sep);
            sb.Append(xs.Last());
            return sb.ToString();
        }

        public static string MakeString<T>(this IEnumerable<T> xs, string start, string sep, string end)
        {
            return String.Format("{0}{1}{2}", start, MakeString(xs, sep), end);
        }
        #endregion
    }


    static class ListFuncsProgram
    {
        public static void Test1()
        {
            IEnumerable<int> xs = Enumerable.Range(1, 10);

            Console.WriteLine("xs: " + xs.MakeString("{", ", ", "}"));
            Console.WriteLine("Head of xs: " + xs.Head());
            Console.WriteLine("Tail of xs: " + xs.Tail().MakeString("{", ", ", "}"));
            Console.WriteLine("Last of xs: " + xs.Last());
            Console.WriteLine("Init of xs: " + xs.Init().MakeString("{", ", ", "}"));
        }
    }
}

File: events/meetup_dotnet_architects_2015/Fun with Functional Programming/F#/ListFuncs.fs
//Sample provided by Fabio Galuppo 
//February 2015 

module ListFuncsProgram

open System

let xs = [1; 1; 2; 3; 5; 8; 13]
let ys = [1..10]
let zs = [for i in 1..10 -> i * i]
let ws = [for (i, j) in (List.zip [1..10] [2.. 2 ..20]) -> i * j]

let head (xs : seq<'a>) : 'a = Seq.head xs

let tail (xs : seq<'a>) : seq<'a> = Seq.skip 1 xs

let init (xs : seq<'a>) : seq<'a> = Seq.take (Math.Max(0, Seq.length xs - 1)) xs

let last (xs : seq<'a>) : 'a = Seq.last xs

let test1() = 
    printfn "xs: %A" xs
    printfn "head of xs: %d" (List.head xs)
    printfn "head of xs: %d" (xs |> List.head)
    printfn "tail of xs: %A" (List.tail xs)
    printfn "tail of xs: %A" (xs |> List.tail)
    printfn "tail of xs: %A" (xs |> Seq.skip 1 |> Seq.toList)
    printfn "last of xs: %d" (xs |> Seq.last)
    printfn "init of xs: %A" (xs |> Seq.take (Math.Max(0, List.length xs - 1)) |> Seq.toList)

let test2() = 
    printfn "xs: %A" xs
    printfn "head of xs: %d" (head xs)
    printfn "tail of xs: %A" (tail xs)
    printfn "last of xs: %d" (last xs)
    printfn "init of xs: %A" (init xs)


